### PR TITLE
Scripts/Stratholme: Baron Rivendare rewrite

### PR DESCRIPTION
**Pull request by Nitair**
**Changes proposed:**

- Match code style
- Fix wrong spell entries `Raise Dead` and `Unholy Aura`
- Fix missing handling of `Unholy Aura` 
- Fix missing emotes during fight (at the relevant parts)
- Changed handling of summoning skeletons, added target location to the DB due `Targets (17, 0) (TARGET_DEST_DB, NO_TARGET)`
- Fix sacrifice of skeletons (`Death Pact`) 

**Target branch:**

- [X] 3.3.5

**Issues addressed:**

- Missing handling of `Unholy Aura`
- Missing handling of sacrificing skeletons (`Death Pact` just heals)
- Missing emotes on spawn of skeletons and on sacrificing skeletons
- Old ScriptDev2 codestyle (also how the skeleton summons are defined)

**Tests performed:**

- [x] Does build on Windows 1803 (Build 17134.1) and on Ubuntu Server 18.04.1
- [x] Tested ingame with level 60 character (some hits on the boss to see the heal)

**Known issues and TO-DO list:**

- [x] `17467 - Unholy Aura is not handled via DB. Don't know what happened and why it's missing
- [x] Add attribute to `17466, 17467 - Unholy Aura` to prevent start of encounter (before pull)
- [x] Add emote: `Text: %s attempts to cast Death Pact on his servants!`
- [x] Add emote: `Text: %s raises an undead servant back to life!`
- [x] Sacrifice doesn't work. The heal is working but the adds doesn't die

** Feedback from reviewers: **
- [x] Remove possible hack in SQL 
>spell_linked_spell in 99% of cases is hack
- [x] Change spell ID of every skeleton summon into enums 
>is better change those magic numbers to enums.

**Notes:**

- any feedback welcome
- build on revision 1daf318cbb8c13d46fb1e207ac54b7f54eafebf3
- continued from #21860

### CODE REVIEW

**Comment by illfated** [comment]
I presume you had a malfunction in the previous PR (**Stratholme rivendare** #22571 ) and had to redo the PR
(linking the previous PR just to show others where it went).

**Comment by Nitair** [comment]
@illfated thanks :)
Yeah it seems that I've selected the master branch to merge which was wrong. ^^ The old PR was #21860 and I wasn't able to get the latest core (just to be sure with the latest changes, because I was longer time away).

File: sql/updates/world/3.3.5/2018_99_99_99_world_335.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+-- Baron Rivendare (Stratholme) - Use skeletons as target
+DELETE FROM `conditions` WHERE `SourceEntry` IN (17698, 17471);
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `Comment`) VALUES 
+(13, 1, 17698, 0, 0, 31, 0, 3, 11197, 0, 0, 0, 0, 'Baron Rivendare (Stratholme) - Death Pact (Heal)'),
+(13, 1, 17471, 0, 0, 31, 0, 3, 11197, 0, 0, 0, 0, 'Baron Rivendare (Stratholme) - Death Pact (Periodic aura)');
+
+-- Baron Rivendare (Stratholme) - Link heal and damage to visual
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger` IN (17698);
+INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES
+(17698, 17471, 'Baron Rivendare (Stratholme) - Link to serverside Death Pact');
```
Comment (Keader) [inline sql/updates/world/3.3.5/2018_99_99_99_world_335.sql]: Awn, please dont do it :/
spell_linked_spell in 99% of cases is hack (well, 100% i guess xD)
Do it using a Spell/AuraScript please <3

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 2]
```diff
@@ -16,170 +15,157 @@
...
+    EVENT_CLEAVE        = 2,
+    EVENT_MORTALSTRIKE  = 3,
+    EVENT_RAISE_DEAD    = 4
+};
 
-class boss_baron_rivendare : public CreatureScript
+// Raise Dead
+uint32 const RaiseDeadSpells[6] =
+{
+    17475, 17476, 17477,
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: is better change those magic numbers to enums.

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 3]
```diff
@@ -16,170 +15,157 @@
...
 public:
-    boss_baron_rivendare() : CreatureScript("boss_baron_rivendare") { }
+    boss_baron_rivendare(Creature* creature) : BossAI(creature, TYPE_BARON), RaiseDead(false) { }
 
-    CreatureAI* GetAI(Creature* creature) const override
+    void Reset() override
     {
-        return GetStratholmeAI<boss_baron_rivendareAI>(creature);
+        // DoCastSelf doesn't add aura after reset
+        me->AddAura(SPELL_UNHOLY_AURA, me);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: Why?
17467 Unholy Aura has a ``` SPELL_AURA_PERIODIC_TRIGGER_SPELL ``` that trigger each 2 secs and is a self cast.
Is better try debug to find why it's not working (if you cant debug, i can help tomorrow)

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 4]
```diff
@@ -16,170 +15,157 @@
...
 
-        uint32 ShadowBolt_Timer;
-        uint32 Cleave_Timer;
-        uint32 MortalStrike_Timer;
-        //    uint32 RaiseDead_Timer;
-        uint32 SummonSkeletons_Timer;
+    void JustDied(Unit* killer) override
+    {
+        // needed until re-write of instance scripts is done
+        instance->SetData(TYPE_BARON, DONE);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: it cant be done inside of InstanceScript?
Call/Handle wherever you want when change boss state?

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 5]
```diff
@@ -16,170 +15,157 @@
...
-            //ShadowBolt
-            if (ShadowBolt_Timer <= diff)
+            switch (eventId)
             {
-                if (SelectTarget(SELECT_TARGET_RANDOM, 0))
-                    DoCastVictim(SPELL_SHADOWBOLT);
+                case EVENT_SHADOWBOLT:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0))
+                        DoCast(target, SPELL_SHADOWBOLT);
+
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: empty space :)

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 6]
```diff
@@ -16,170 +15,157 @@
...
+                    break;
+                case EVENT_MORTALSTRIKE:
+                    DoCastVictim(SPELL_MORTALSTRIKE);
+                    events.Repeat(10s, 25s);
+                    break;
+                case EVENT_RAISE_DEAD:
+                    if (!RaiseDead)
+                    {
+                        DoCast(SPELL_RAISE_DEAD);
+
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: not needed space

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 7]
```diff
@@ -16,170 +15,157 @@
...
+                    events.Repeat(7s, 17s);
+                    break;
+                case EVENT_MORTALSTRIKE:
+                    DoCastVictim(SPELL_MORTALSTRIKE);
+                    events.Repeat(10s, 25s);
+                    break;
+                case EVENT_RAISE_DEAD:
+                    if (!RaiseDead)
+                    {
+                        DoCast(SPELL_RAISE_DEAD);
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: DoCastSelf please.
DoCast will force core chose a target, and this case target will be caster, so is better explicit it :)

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 8]
```diff
@@ -16,170 +15,157 @@
...
+                        Talk(EMOTE_RAISE_DEAD);
+                    }
+                    else
+                    {
+                        RaiseDead = false;
+                        Talk(EMOTE_DEATH_PACT);
+                    }
+                    events.Repeat(12s);
+                    break;
+                }
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: missing default: xD

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 9]
```diff
@@ -16,170 +15,157 @@
...
-
-            DoMeleeAttackIfReady();
-        }
-    };
+// Death Pact 3 (17472) needs to be casted by the skeletons
+struct npc_summoned_skeleton : public ScriptedAI
+{
+    npc_summoned_skeleton(Creature* creature) : ScriptedAI(creature) { }
 
+    void SpellHit(Unit* /*caster*/, SpellInfo const* spell) override
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: It's not wrong... but looks like a job to a spell script handle with it

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp (hunk 5 above)
Comment (illfated) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: yup, redundant blank line.

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp (hunk 6 above)
Comment (illfated) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: isn't it OK to have a blank line before if/for loops?
... or maybe it deviates from TC code style.

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp (hunk 3 above)
Comment (Nitair) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: Yeah sure would be great.

**Comment by jackpoz** [comment]
@Nitair could you apply the changes requested by @Keader ?

**Comment by Nitair** [comment]
@jackpoz - I will do that, but at the spellscript part is something I've never worked with. I will need some time learn how to use that properly.

**Comment by jackpoz** [comment]
maybe @Keader can help with that part

**Comment by Nitair** [comment]
I have something like (now you can imagine, that I've no clue lol). I cannot test that yet, because I've dropped my test environment.
```
class spell_rivendare_death_pact_2 : public SpellScript
{
    PrepareSpellScript(spell_rivendare_death_pact_2);

    void OnHit(SpellEffIndex effIndex)
    {
        GetCaster()->GetVictim()->CastSpell(0, SPELL_DEATH_PACT_3, true);
    }

    void Register() override
    {
        OnEffectHitTarget += SpellEffectFn(spell_rivendare_death_pact_2::OnHit, EFFECT_0, SPELL_SCRIPT_HOOK_EFFECT_HIT);
    }
};
```

**Comment by Nitair** [comment]
Okay I got some free time in the next days. I will try to finish this script.

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2008-2019 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: revert xd

Comment (Nitair) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: lol

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 11]
```diff
@@ -98,10 +98,11 @@ struct boss_baron_rivendare : public BossAI
     void UpdateAI(uint32 diff) override
     {
         if (!UpdateVictim())
+        {
```
Comment (Nitair) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: Could someone check if this is valid?

**Comment by illfated** [comment]
Not wanting to draw attention away from your question about the code section, but maybe you could rebase your PR branch on top of up-to-date 3.3.5 branch to make it mergeable, so AppVeyor and travis-ci can run again?

**Comment by Nitair** [comment]
@illfated Happy? :)

**Comment by illfated** [comment]
Thank you very much, nice to see that the checks don't complain when the branch is up to date.

BTW, you asked
> Could someone check if this is valid?

Did you mean the whole section where you placed the question
```cpp
    void UpdateAI(uint32 diff) override
    {
        if (!UpdateVictim())
        {
            if (!me->HasAura(SPELL_UNHOLY_AURA))
                DoCastSelf(SPELL_UNHOLY_AURA);
            return;
        }
```
... or did you mean only the `if (!UpdateVictim())` line?

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp [hunk 12]
```diff
@@ -16,170 +15,160 @@
...
+    {
+        if (!UpdateVictim())
         {
-            //can't use entercombat(), boss' dmg aura sets near players in combat, before entering the room's door
-            if (instance->GetData(TYPE_BARON) == NOT_STARTED)
-                instance->SetData(TYPE_BARON, IN_PROGRESS);
-            ScriptedAI::AttackStart(who);
+            if (!me->HasAura(SPELL_UNHOLY_AURA))
+                DoCastSelf(SPELL_UNHOLY_AURA);
+            return;
```
Comment (Killyana) [inline src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: Put it in creature_template_addon

Comment (illfated) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: Like this one for Horgus the Ravager ?
```sql
SELECT * FROM `creature_template_addon` WHERE `auras`= 17467;
```
```sql
entry  path_id   mount  bytes1  bytes2   emote  visibilityDistanceType  auras
-----  -------  ------  ------  ------  ------  ----------------------  -----
10946        0       0       0       0       0                       0  17467
```

Comment (Nitair) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: @Killyana , as far as I remember this was the first thing I've tried. The aura doesn't apply.

Edit: I've checked again and it doesn't apply.

Comment (Killyana) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: It works for me
``` sql
DELETE FROM `creature_addon` WHERE `guid` IN(54241);
UPDATE `creature_template_addon` SET `auras`='17467' WHERE `entry` IN (10440);

Comment (Nitair) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: @Killyana Uh ```DELETE FROM `creature_addon` WHERE `guid` IN(54241);``` is something I never did. Now its working fine. Will update the PR.

Thanks a lot <3

**Comment by Nitair** [comment]
for the instancescript part -> I'm already on it, but I will let this rivendare script with the old handling from sd2

I will create a PR for that, after I've managed to get it close to the TC-way... I guess it will need some changes as well, as every PR from my side :D

**Comment by Nitair** [comment]
@illfated my question is already answered by @Killyana :)

**Comment by jackpoz** [comment]
what's the status of this PR ? is it ready to be merged or are there still pending things ?

**Comment by illfated** [comment]
Judging by the PR comment on top, it looks like all the boxes are ticked, as well as nothing wrong in the tests. Unless someone can find any remaining issues by reading through the code, a reply from @Nitair would be good.

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp (hunk 4 above)
Comment (illfated) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: @Nitair : Please post a reply to this question.

**Comment by Nitair** [comment]
@jackpoz @illfated Should be all done from my end. :)

**Comment by illfated** [comment]
Thank you for replying. Could you also post a comment reply to Keader in https://github.com/TrinityCore/TrinityCore/pull/22572#discussion_r224459640 ?

File: src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp (hunk 4 above)
Comment (Nitair) [reply src/server/scripts/EasternKingdoms/Stratholme/boss_baron_rivendare.cpp]: @Keader @illfated

as written https://github.com/TrinityCore/TrinityCore/pull/21687 it cant be done until the instance script is re-written. I'm already on it but it needs some time until creating the PR here (as written https://github.com/TrinityCore/TrinityCore/pull/22572#issuecomment-480497510)

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**